### DISCUSSION TITLE
How to use debug? (--debug-asset, --debug-similarity, etc.)

### DISCUSSION THREAD

**Discussion by Synexcu** [XVs32/kcauto_custom#72] [Q&A] [answered]
Hello, it's me again, sorry if I come quite often.

As the title says, How do you use debug? I've read the code from `.\util\debug.py` and in my understanding it is used to find coordinates where an asset is located in the game screen. I tried doing it with `python kcauto --debug --debug-asset "combat_continue.png"` but it cant run it.

it comes with this error:
```
File "kcauto\util\debug.py", line 15, in find_all
    results = region.find_all(asset, similarity)
  File "C:\Users\AppData\Local\Programs\Python\Python37\lib\site-packages\pyvisauto\__init__.py", line 136, in find_all
    matches = self._match_template(target, template, cached=cached)
  File "C:\Users\AppData\Local\Programs\Python\Python37\lib\site-packages\pyvisauto\__init__.py", line 70, in _match_template
    template.shape[::-1]
AttributeError: 'NoneType' object has no attribute 'shape'
```

I need to use the debug so that I can make a new logic where you can force push if you have damecon in your ship (if your FS is taiha).

I have also seen one of your enhancement discussion where you mentioned you wanted to make a CLI  debug feature, but I don't know if you have made one or not..

Thanks again!

**Comment by XVs32** [comment, marked as answer]
Actually I have never used the `--debug-asset` and `--debug-similarity` function.
Usually I would just throw a `kca_u.kca.exists("range", "pic_name.png")` to see if an asset exist.

What I mean for the CLI debug feature is adding '--debug-output' option there and it just control the log level (prints `Log.log_debug()` too),
and .... sadly that is not urgent enough to come up to my next to-do list.
(But sure, you can simply use `python kcauto --debug-output in command line for the log`)

### Combat node detect is not reliable (enough) at the moment
What I would want to mention though, is the current combat logic is running base on the APIs from kancolle,
and the original kcauto uses single thread to listen to the APIs and to control the mouse.
This could sometimes lead to mis-detecting the current node your fleet is standing on [ref](https://github.com/XVs32/kcauto_custom/issues/63#issuecomment-1700822110) .
Obviously this would affect the force push logic you are attempting to add.

Although I said that I am not going to redo the combat logic in the previous post,
but this mis-detection is also affecting node-selection/formation-selection etc. and it is getting critical.
So I am currently working on the combat logic redo (separate the API listener to another thread).

### You might have to tweak you code in the future if you want to merge this fix to your code.

**Comment by Synexcu** [comment]
Say for example, you would want to debug the `"combat_continue.png"` asset (`kcauto\assets\combat\combat_continue.png`), what kind of argument that you have to put in into the cmd prompt?

Also, what I meant in pushing where you have damecon in your FS and your FS is Taiha, is like the following:

In here, your FS is taiha, and thus you can still continue the sortie **if** your FS has a damecon/damegami equipped instead of forcibly withdrawn by the game:

![image](https://github.com/XVs32/kcauto_custom/assets/82300898/b9401799-0a3d-432b-aa88-6d406a9a3f29)

In your code in your previous [discussion](https://github.com/XVs32/kcauto_custom/discussions/62#discussioncomment-6865673), from my understanding, you *will* continue the sortie, **but** the code underneath it (in `.\combat\combat_core.py`) prevents you to continue forward if you got the prompt I have shown you earlier.

```

if continue_sortie:
            Log.log_msg("Continuing sortie.")

>             kca_u.kca.click_existing(
>                 'kc', 'combat|combat_continue.png', cached=True)

        else:
            Log.log_msg("Retreating from sortie.")
            kca_u.kca.click_existing(
                'kc', 'combat|combat_retreat.png', cached=True)
        kca_u.kca.r['lbas'].hover()

        return continue_sortie

```


Instead, I have created my own asset called `combat_continue_dame.png` in which then will be used to scan whether the button in the prompt above is present in the game screen and then proceeds to click it if it is there.

![combat_continue_dame](https://github.com/XVs32/kcauto_custom/assets/82300898/cf502591-17c9-4cde-8307-3e432bb53172)

```
        if continue_sortie:
            Log.log_msg("Continuing sortie.")

>             if kca_u.kca.exists('kc', 'combat|combat_continue_dame.png'):
>                 kca_u.kca.click_existing(
>                     'kc', 'combat|combat_continue_dame.png', cached=True)

            else:
                kca_u.kca.click_existing(
                    'kc', 'combat|combat_continue.png', cached=True)
        else:
            Log.log_msg("Retreating from sortie.")
            kca_u.kca.click_existing(
                'kc', 'combat|combat_retreat.png', cached=True)
        kca_u.kca.r['lbas'].hover()

        return continue_sortie
```

But here's what I'm confused in, I don't know if the code I have inserted will fully functionally work, because I don't know if kcauto could see it and I don't know how I can see what kcauto sees, like, to see the computer vision.

In `kca_u.kca.click_existing('kc', 'combat|combat_continue_dame.png', cached=True)`, I still don't quite understand how the code works. What is `cached=True` mean? What is the `kc` refers to? Is it in ` .\util\kca.py`? Because the only `kc` I see is `self._create_or_shift_region('kc', x, y, w, h)` and I'm not quite sure it is right and what it actually is..

Also, if by any chance if code *is* correct, do I need to make a new coordinate to click the button in the prompt? Because from the code in `.\util\kca.py`, `click_existing` creates its own pinpoint on where to click according to where the asset is located in the game screen, but again, I still don't know because I don't have any way to verify my assumption..

Oh, and sorry.. I haven't learned about threads.. So there's very few things I may or may not understand within Threads topic..


Thanks!

**Comment by XVs32** [comment]
### To see if an asset appear in game screen.
1. Use `kca_u.kca.exists()` function, it returns `true` if the specified asset exists.  
For example, you could simply throw a `print` in your `if` cause
```
        if continue_sortie:
            Log.log_msg("Continuing sortie.")

>             if kca_u.kca.exists('kc', 'combat|combat_continue_dame.png'):
>>               print("PIC found")
>                 kca_u.kca.click_existing(
>                     'kc', 'combat|combat_continue_dame.png', cached=True)
```
2. Write a simple standalone .py to do this.  
For example, I wrote `./tool/pic_find.py` to compare `a.png` and `b.png`, and the .py tells you if `b.png` is a sub-image of `a.png`.

---

### Usage of `kca_u.kca.click_existing()`
The original kcauto author (the first author of kcauto, I believe) did a great job on commenting,  
there is explanation for (almost) all the util functions.
You can find it in the function declaration.

For example:
```
    def click_existing(
            self, region, asset, similarity=DEFAULT, pad=(0, 0, 0, 0),
            cached=False):
        """Helper method that clicks a region if it exists. The pad parameter
        allows for further tweaking of the valid click region.

        Args:
            region (Region, Match, str): Region/Match object or pre-defined
                region key.
            asset (str): kcauto-style asset path.
            similarity (float, optional): minimum similarity score. Defaults to
                DEFAULT.
            pad (tuple, optional): click region modifier. Defaults to
                (0, 0, 0, 0).
            cached (bool, optional): flag for using cached snapshot of region.
                Defaults to False.

        Returns:
            bool: True if the region was found and clicked; False otherwise.
        """
        r = self._get_region(region)
...
...
```

Well, I could see it is still kinda hard to understand the comment, so I will give a brief explanation here:

1. region: This is the search region, it tells kcauto where to search,  
For example the `kc` is `x, y, w, h`, `x` and `y` are the top left of the game screen, `w` and `h` are the width and height of the game screen,  
Therefore, `kc` is the whole game screen.  
We have many more regions like `left`, `right` etc. And usually you would want to use the smallest region possible.  
Cause believe it or not, it takes a lot computing power to do a screen search.

2. pad: The region to click, kcauto does not always click on the same point (to prevent anti bot detection),  
And `pad` is the range kacuto could click.  
If `pad` is (0,0,0,0), kcauto will click on the top left of the asset found (not recommend).

3. cached: Besides of screen search, taking a screen shot is also a heavy lifting for the PC,  
When `cached` is enabled, kcauto does not take a new screen shot for the current function.

---

PS. Don't worry about the multi-thread thing, I will handle that.  
It was just a kind reminder to inform you that the kcauto combat logic is changing soon,  
and you might have to update your code when it happens.

**Reply by Synexcu** [reply]
Woah I was just about to write an update on this haha, your comment surprised me there

**Comment by Synexcu** [comment]
So, Update:

I finally made it work. After reading the code, it does make its own coordinate automatically based on the location where our asset is present in the game screen. And.. It works! I guess..

So, I made 2 new assets:
> `combat_continue_damecon` for damecon button

![combat_continue_damecon](https://github.com/XVs32/kcauto_custom/assets/82300898/f5b7d53c-8c16-497e-ad97-49250a69c9fa)

> `combat_continue_dame` for damegami button

![combat_continue_dame](https://github.com/XVs32/kcauto_custom/assets/82300898/a8d221e9-3fa8-48a2-967a-2c9e6137c2dd)


And then I made several edits at the code:

> `.\combat_core.py`


```
..
..
                    Log.log_debug("Check for Port API.")
                    for ship_idx, ship in enumerate(flt.fleets.combat_ships):
                        Log.log_debug(str(ship.name) + ": " + str(ship.damage.display_name) + " damage")
                        if ship.damage is DamageStateEnum.HEAVY:
                            Log.log_debug("Flagship heavily damaged.")

>                             if self._resolve_continue_sortie_prompt():
>                                 Log.log_msg("You passed this checkpoint!")
>                                 break
>                             else:

                                self._click_until_port()

>                         Log.log_msg("You came back")

                        break
...
...
...
...
...
        if continue_sortie:
            Log.log_msg("Continuing sortie.")

>             if kca_u.kca.exists('kc', 'combat|combat_continue_damecon.png'):
>                 kca_u.kca.click_existing(
>                     'kc', 'combat|combat_continue_damecon.png', cached=True)
>             elif kca_u.kca.exists('kc', 'combat|combat_continue_dame.png'):
>                 kca_u.kca.click_existing(
>                     'kc', 'combat|combat_continue_dame.png', cached=True)
>             else:

                kca_u.kca.click_existing(
                    'kc', 'combat|combat_continue.png', cached=True)

>             Log.log_msg("Damecon CHECK!")

        else:
            Log.log_msg("Retreating from sortie.")
            kca_u.kca.click_existing(
                'kc', 'combat|combat_retreat.png', cached=True)
        kca_u.kca.r['lbas'].hover()
```


It works, but the thing is, it kinda 'loops' the `continue_sortie` part because this is what I got in the debug log:

```
[2023-12-30 02:19:39] Processing battle results.
[2023-12-30 02:19:39] Battle animations complete.
[2023-12-30 02:19:40] Loading data from 'gui_intervals.json'.
[2023-12-30 02:19:41] Running visual health check of fleet 1.
[2023-12-30 02:19:41] Visual health check results: {'heavy': 1, 'moderate': 0, 'minor': 0}
[2023-12-30 02:19:41] Fleet 1 / Heavy fleet damage / No Fatigue
[2023-12-30 02:19:41] Amatsukaze (Heavy damage)

[2023-12-30 02:19:41] Check for Port API.

[2023-12-30 02:19:41] Amatsukaze: Heavy damage
[2023-12-30 02:19:41] Flagship heavily damaged.

[2023-12-30 02:19:41] Resolve continue sortie prompt.             <- Start of `continue_sortie`
[2023-12-30 02:19:41] A is specified as a push node.
[2023-12-30 02:19:41] Continuing sortie.
[2023-12-30 02:19:41] Damecon CHECK!                              <- 1st Damecon checkpoint here
[2023-12-30 02:19:42] You passed this checkpoint!

[2023-12-30 02:19:42] Begin waiting for API payload(s).

[2023-12-30 02:19:46] Check for Port API.

[2023-12-30 02:19:46] Amatsukaze: Heavy damage
[2023-12-30 02:19:46] Flagship heavily damaged.

[2023-12-30 02:19:46] Resolve continue sortie prompt.             <- Start of `continue_sortie`
[2023-12-30 02:19:46] A is specified as a push node.
[2023-12-30 02:19:46] Continuing sortie.
[2023-12-30 02:19:46] Damecon CHECK!                              <- 2nd Damecon checkpoint here
[2023-12-30 02:19:46] You passed this checkpoint!

[2023-12-30 02:19:46] Begin waiting for API payload(s).

[2023-12-30 02:19:50] Continue sortie prompt.

[2023-12-30 02:19:50] Resolve continue sortie prompt.             <- Start of `continue_sortie`
[2023-12-30 02:19:50] A is specified as a push node.
[2023-12-30 02:19:50] Continuing sortie.
[2023-12-30 02:19:51] Loading data from 'gui_intervals.json'.
[2023-12-30 02:19:51] Damecon CHECK!                              <- 3rd Damecon checkpoint here

[2023-12-30 02:19:52] Begin waiting for API payload(s).
[2023-12-30 02:19:52] Waiting for request X (idk if it is safe to paste here but I'll just blank the request out)
[2023-12-30 02:19:52] Request X received
[2023-12-30 02:19:52] Updating ship data from API.
[2023-12-30 02:19:52] Between nodes.
[2023-12-30 02:19:52] Wait for combat API.

[2023-12-30 02:19:52] Begin waiting for API payload(s).
[2023-12-30 02:19:52] Waiting for request Y
[2023-12-30 02:19:52] Request Y received
[2023-12-30 02:19:52] Going to Node B
```

I don't know what's causing it to iterate it 3 times. It works, and it did click onto the designated asset area on the game screen and successfully continued the fleet to 3-2 ammo node, but the log showed that kcauto ran `continue_sortie` 3 times, which in my opinion is concerning because maybe some issue may arise in the future, and since it is connected to the `_resolve_continue_sortie_prompt()` it is dangerous to leave it as it is right now..

Oh yeah, I also read your comment on `pad`:

> pad: The region to click, kcauto does not always click on the same point (to prevent anti bot detection),
> And `pad` is the range kacuto could click.
> If `pad` is (0,0,0,0), kcauto will click on the top left of the asset found (not recommend).

Could you explain why (0,0,0,0) pad is not recommended? Is it safe to change the pad values? And if so, how do I change the `pad` and what value would you recommend?


Thanks a much!

**Comment by XVs32** [comment]
1. _resolve_continue_sortie_prompt() does handle the clicking of continue/retreat button,  
normally it should runs ONE time only, but now you put it in a for loop, so it runs for more than one time,  
I think that's the reason why that "loop" occurred

2. As I said, `pad` is the region where kcauto is allowed to click,  
if you set it to (0,0,0,0), kcauto will have no choice but to click on the same spot(pixel) every time, which is no good against bot detection
For the range for `pad`, it depends. You could have a bigger `pad` if the button is huge, or when button is tiny, you just have to stick with a small `pad`

**Comment by Synexcu** [comment]
Oooh found it. Thanks for the answers!
